### PR TITLE
Honor useSpellingFunctionality parameter when reporting shortcuts

### LINKED ISSUE #15566: Updated shortcut key reporting doesn't respect use spelling functionality option
### Steps to reproduce:
1. Disable the use spelling functionality option.
2. Navigate to a control with a shortcut key, for example by pressing alt+f in Notepad.
### Actual behavior:

Character mode commands are still sent to the synthesizer. This can cause issues with some older SAPI synthesizers that don't properly implement character mode support.
### Expected behavior:

No character mode commands should be sent to the synthesizer.
### NVDA logs, crash dumps and other attachments:
Log fragment with use spelling functionality off.
IO - speech.speech.speak (08:29:13.122) - MainThread (20000):
Speaking [LangChangeCommand ('en_US'), 'Installed apps', 'Alt, ', CharacterModeCommand(True), 'p', CharacterModeCommand(False), '1 of 18', CancellableSpeech (still valid)]
#### NVDA installed/portable/running from source:
Installed.
#### NVDA version:
alpha-29490,2adb77cb.
#### Windows version:
Windows 11, Version 23H2 (OS Build 25951.1010).
#### Name and version of other software in use when reproducing the issue:
N/A.
#### Other information about your system:
N/A.
#### Does the issue still occur after restarting your computer?
Yes.
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
Yes, versions prior to #14900 being merged don't have this problem.
#### If NVDA add-ons are disabled, is your problem still occurring?
Yes.
#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Yes.

### PR DESCRIPTION
**Pull request by CyrilleB79**
### Link to issue number:
Closes #15566
Fix-up of #14900 / #15373.
### Summary of the issue:
In #14900 / #15373., spelling mode was used (`CharacterModeCommand`) no matter the state of the "Use spelling functionality" option.

The use spelling functionality exists because some old SAPI synth do not support spelling mode correctly. For them, this option needs to be honored.

### Description of user facing changes
Use spelling functionality option will now be honored when reporting shortcuts.
### Description of development approach
See logic in the code.
If use spelling functionality option is disabled, the key is just reported as is.

### Testing strategy:
Manual check:
* Checked that reporting shortcuts is still OK
* Checked in the log that `CharacterModeCommand` is used only when the "Use spelling functionality" option is enabled.
### Known issues with pull request:
None

### Change log

None: unreleased issue.

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

### CODE REVIEW

**Comment by CyrilleB79** [comment]
@seanbudd or any other reviewer:
I have had to fix the unit tests. Not sure however that function patching is the correct way to do this. Let me know if another approach is required.

File: tests/unit/test_speechShortcutKeys.py [hunk 1]
```diff
@@ -12,11 +12,22 @@
...
+import speech.shortcutKeys
 from speech.commands import CharacterModeCommand
 
 
+original_shouldUseSpellingFunctionality = speech.shortcutKeys.shouldUseSpellingFunctionality
+
 class Test_getKeyboardShortcutSpeech(unittest.TestCase):
 
+	@classmethod
+	def setUpClass(cls):
```
Comment (lukaszgo1) [inline tests/unit/test_speechShortcutKeys.py:24]: Have you considered using `unittest.mock.patch` for replacing the function during the tests? When used as a decorator on these test classes you'll have much less of boilerplate code.

**Review by lukaszgo1 — COMMENTED** [review]
Pretty nice overall!

**Comment by CyrilleB79** [comment]
@lukaszgo1, many thanks for the review!

I am not used to `unittest.mock.patch` (and more generally nor to `mock`), but that's a really nice tool. Could you please double check that I am using it correctly?

**Comment by CyrilleB79** [comment]
@lukaszgo1, I think you are a user of MathPlayer.
Could you manual test the MathPlayer code that has been modified in this PR?

File: source/speech/shortcutKeys.py [hunk 2]
```diff
@@ -79,6 +81,10 @@ def _getKeyboardShortcutSpeech(keyboardShortcut: str) -> SpeechSequence:
 	return seq
 
 
+def shouldUseSpellingFunctionality() -> bool:
+	synth = getSynth()
+	return config.conf["speech"][synth.name]["useSpellingFunctionality"]
+
```
Comment (seanbudd) [inline source/speech/shortcutKeys.py:87]: Fix lint

```suggestion


```

File: tests/unit/test_speechShortcutKeys.py [hunk 3]
```diff
@@ -30,6 +33,21 @@ def test_simpleLetterKey(self):
 		)
 		self.assertEqual(repr(output), expected)
 
+	@patch('speech.shortcutKeys.shouldUseSpellingFunctionality', lambda: False)
+	def test_simpleLetterKeyWithSpellingFunctionalityDisabled(self):
+		"""A shortcut consisting in only one letter in the case where "Use spelling functionality" is disabled
+		(see #15566).
+		."""
```
Comment (seanbudd) [inline tests/unit/test_speechShortcutKeys.py:40]:
```suggestion
		"""
```

File: tests/unit/test_speechShortcutKeys.py [hunk 4]
```diff
@@ -30,6 +33,21 @@ def test_simpleLetterKey(self):
...
+
+		expected = repr([
+			'A',
+		])
```
Comment (seanbudd) [inline tests/unit/test_speechShortcutKeys.py:44]:
```suggestion
		expected = repr(['A',])
```

**Review by seanbudd — APPROVED** [review]
Thanks @CyrilleB79

File: tests/unit/test_speechShortcutKeys.py [hunk 5]
```diff
@@ -30,6 +33,19 @@ def test_simpleLetterKey(self):
 		)
 		self.assertEqual(repr(output), expected)
 
+	@patch('speech.shortcutKeys.shouldUseSpellingFunctionality', lambda: False)
+	def test_simpleLetterKeyWithSpellingFunctionalityDisabled(self):
+		"""A shortcut consisting in only one letter in the case where "Use spelling functionality" is disabled
+		(see #15566).
+		"""
+
+		expected = repr(['A',])
```
Comment (seanbudd) [inline tests/unit/test_speechShortcutKeys.py:42]:
```suggestion
		expected = repr(['A', ])
```

**Pull request merged**